tensorflow/c/experimental/grappler/grappler.cc

```diff
@@ -0,0 +1,194 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+// This file extends/implements core graph optimizer base classes in terms of
+// the C API defined in grappler.h. A class "CSomething" represents a
+// "Something" that can be manipulated via calls in the C interface and a C
+// struct called "TP_Something".
+
+#include "tensorflow/c/experimental/grappler/grappler.h"
+
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "tensorflow/c/c_api_internal.h"
+#include "tensorflow/c/experimental/grappler/grappler_internal.h"
+#include "tensorflow/c/tf_status_helper.h"
+#include "tensorflow/core/framework/function.h"
+#include "tensorflow/core/grappler/costs/graph_properties.h"
+#include "tensorflow/core/grappler/grappler_item.h"
+#include "tensorflow/core/platform/errors.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/status.h"
+
+namespace {
+
+#define VALIDATE_STRUCT_SIZE(STRUCT_NAME, STRUCT_OBJ, SIZE_VALUE_NAME)    \
+  do {                                                                    \
+    if (STRUCT_OBJ.struct_size == 0) {                                    \
+      return tensorflow::Status(tensorflow::error::FAILED_PRECONDITION,   \
+                                "struct_size field in " #STRUCT_NAME      \
+                                " must be set to " #SIZE_VALUE_NAME "."); \
+    }                                                                     \
+  } while (0)
+
+#define VALIDATE_MEMBER(STRUCT_NAME, STRUCT_OBJ, NAME)                  \
+  do {                                                                  \
+    if (STRUCT_OBJ.NAME == 0) {                                         \
+      return tensorflow::Status(tensorflow::error::FAILED_PRECONDITION, \
+                                "'" #NAME "' field in " #STRUCT_NAME    \
+                                " must be set.");                       \
+    }                                                                   \
+  } while (0)
+
+tensorflow::Status ValidateTPOptimizerRegistrationParams(
+    const TP_OptimizerRegistrationParams& params) {
+  VALIDATE_STRUCT_SIZE(TP_OptimizerRegistrationParams, params,
+                       TP_OPTIMIZER_REGISTRARION_PARAMS_STRUCT_SIZE);
```

Comment:
Please fix in other places as well.
```suggestion
                       TP_OPTIMIZER_REGISTRATION_PARAMS_STRUCT_SIZE);
```

Reply:
Done.